## Math 3790 W01
## Yuge Xu 1194170
## NLP Method: T-test

In [1]:
import nltk
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.probability import FreqDist
from nltk import bigrams
import pandas as pd
import numpy as np
from nltk import pos_tag
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

In [2]:
# Load Data
with open("Corpus_Country_History.txt", "r", encoding="utf-8") as file:
    text = file.read()
text = text.replace('\n', ' ')
words = word_tokenize(text)
words = [word for word in words if word.isalpha()]
sentences = sent_tokenize(text)
print(f"len(sentences)= {len(sentences)} ")
print(f"len(words) = {len(words)}")

bigram_list = list(bigrams(words))
# filter the bigram lists by cleaning some stop words
stop_words = set(stopwords.words('english'))
bigram_list = [bigram for bigram in bigram_list if not any(word.lower() in stop_words for word in bigram)]



len(sentences)= 37715 
len(words) = 755386


In [3]:
# find target bigram, and count how many times the target bigram occurrs
def find_bigram(bigram_list, target_bigram):
    count = 0
    # Converts all the words in the target binary to lowercase
    target_bigram_lower = tuple(word.lower() for word in target_bigram)
    for bigram in bigram_list:
        bigram_lower = tuple(word.lower() for word in bigram)
        if bigram_lower == target_bigram_lower:
            count += 1
    return count

# count the number of times a keyword appears in whole bigram lists
def count_keyword_in_bigrams(keyword, bigram_list):
    count = 0
    keyword = keyword.lower()  
    for bigram in bigram_list:
        bigram = tuple(word.lower() for word in bigram)
        if keyword in bigram:
            count += 1
    return count
# define a method for calculating t-score
def calculate_t_score(x_bar,mu,S_square,N):
    return (x_bar - mu) / np.sqrt(S_square / N)

In [4]:
def get_t_score_report(w1,w2):
    # w1 = 'world'
    # w2 = 'inequality'
    target_bigram = (w1, w2)
    count = find_bigram(bigram_list, target_bigram)
    num_of_bigram_contains_w1_w2 = count
    print(f"The bigram '{target_bigram}' appears {count} times in the bigram list.")
    num_of_bigram_contains_w1 = count_keyword_in_bigrams(w1, bigram_list)
    num_of_bigram_contains_w2 = count_keyword_in_bigrams(w2, bigram_list)
    total_num_of_bigram = len(words)-1
    print(f"num_of_bigram_contains_w1 {w1} =  {num_of_bigram_contains_w1} ")
    print(f"num_of_bigram_contains_w2 {w2} = {num_of_bigram_contains_w2} ")
    print(f"num_of_bigram_contains_w1_w2 {target_bigram} = {num_of_bigram_contains_w1_w2} ")
    print(f"total_num_of_bigram =  {total_num_of_bigram} ")

    probability_w1 = num_of_bigram_contains_w1/total_num_of_bigram
    probability_w2 = num_of_bigram_contains_w2/total_num_of_bigram
    print(f"probability_w1 {w1} = {probability_w1}")
    print(f"probability_w2 {w2} = {probability_w2}")

    h0_probability_w1_w2 = probability_w1*probability_w2 # null hypothesis, independent probability
    print(f"null hypothesis h0 probability = {h0_probability_w1_w2}")
    x_bar = num_of_bigram_contains_w1_w2/total_num_of_bigram
    print(f"sample mean = {x_bar}")
    mu = h0_probability_w1_w2
    s_square = num_of_bigram_contains_w1_w2/total_num_of_bigram
    t_score = calculate_t_score(x_bar,mu,s_square,total_num_of_bigram)
    print(f"t-score for bigram {target_bigram} = {t_score}")


In [5]:
get_t_score_report('world','war')

The bigram '('world', 'war')' appears 166 times in the bigram list.
num_of_bigram_contains_w1 world =  812 
num_of_bigram_contains_w2 war = 1316 
num_of_bigram_contains_w1_w2 ('world', 'war') = 166 
total_num_of_bigram =  755385 
probability_w1 world = 0.0010749485361769165
probability_w2 war = 0.0017421579724246575
null hypothesis h0 probability = 1.8727301622468305e-06
sample mean = 0.00021975548892286714
t-score for bigram ('world', 'war') = 12.774301968440861


In [6]:
get_t_score_report('world','movement')

The bigram '('world', 'movement')' appears 1 times in the bigram list.
num_of_bigram_contains_w1 world =  812 
num_of_bigram_contains_w2 movement = 285 
num_of_bigram_contains_w1_w2 ('world', 'movement') = 1 
total_num_of_bigram =  755385 
probability_w1 world = 0.0010749485361769165
probability_w2 movement = 0.0003772910502591394
null hypothesis h0 probability = 4.055684621887133e-07
sample mean = 1.323828246523296e-06
t-score for bigram ('world', 'movement') = 0.6936396671895787


In [7]:
get_t_score_report('world','home')

The bigram '('world', 'home')' appears 1 times in the bigram list.
num_of_bigram_contains_w1 world =  812 
num_of_bigram_contains_w2 home = 185 
num_of_bigram_contains_w1_w2 ('world', 'home') = 1 
total_num_of_bigram =  755385 
probability_w1 world = 0.0010749485361769165
probability_w2 home = 0.00024490822560680977
null hypothesis h0 probability = 2.632637386137262e-07
sample mean = 1.323828246523296e-06
t-score for bigram ('world', 'home') = 0.8011345208072704


In [8]:
get_t_score_report('world','bank')

The bigram '('world', 'bank')' appears 9 times in the bigram list.
num_of_bigram_contains_w1 world =  812 
num_of_bigram_contains_w2 bank = 142 
num_of_bigram_contains_w1_w2 ('world', 'bank') = 9 
total_num_of_bigram =  755385 
probability_w1 world = 0.0010749485361769165
probability_w2 bank = 0.00018798361100630804
null hypothesis h0 probability = 2.0207270747648172e-07
sample mean = 1.1914454218709665e-05
t-score for bigram ('world', 'bank') = 2.9491191026209593


|t-score|collocation|
|---|---|
|0.6936396671895787|(world,movement)|
|0.8011345208072704|(world,home)|
|2.061081675157631|(world,population)|
|2.1134815818236437|(world,market)|
|2.1711692115212418|(world,inequality)|
|2.9491191026209593|(world,bank)|
|12.77430211379272|(world,war)|